proxy/envoy/resources.go

```diff
@@ -230,6 +230,34 @@ type TCPRoute struct {
 	DestinationPorts  string   `json:"destination_ports,omitempty"`
 	SourceIPList      []string `json:"source_ip_list,omitempty"`
 	SourcePorts       string   `json:"source_ports,omitempty"`
+
+	// special value to retain dependent cluster definition for TCP routes.
+	clusterRef *Cluster
+}
+
+// Merge operation selects a union of two route configs prioritizing the first.
+func (rc *TCPRouteConfig) merge(that *TCPRouteConfig) *TCPRouteConfig {
```

Comment:
We have access to destination service hostname via TCPRoute.clusterRef.hostname, but I'm not sure how they would be used to merge with HTTP since envoy manages listeners by port?

re: preferring inbound over outbound, I thought that would be handled by lines 136-149 which merges inbound/outbound routes which prefers inbound over outbound. Did you have something else in mind?